### PR TITLE
[ShellScript] Improve built-in and compound test expressions

### PR DESCRIPTION
**Pull request by deathaxe**
Addresses #4272 / no bare glob qualifiers in compound tests

This PR refactors built-in `test ...` and `[ ... ]` as well as compound `[[ ... ]]` test expressions to apply more detailed parsing rules, which follow more closely closely to language specifications and shell parsing behavior.

This fixes various edge cases and enables more fine gained and accurate highlighting.

Parsing performance isn't affected by this change, syntax cache size is also nearly unchanged.

For details, please refer to commit messages.

Some highlights:

- highlight math expressions before and after `-eq` and friends

  grafik
   
- support some of Bash/Zsh's insanity

  grafik

- compare two redirections

  grafik

- fix arbitrarily nested groups in expressions

  grafik

  grafik

- no bare globs in compound tests

  grafik

  grafik

- various other fixes

### CODE REVIEW

**Review by keith-hall — DISMISSED** [review]

**Comment by deathaxe** [comment]
Most recent commit "just" creates dedicated section for compound tests, renames contexts and includes compound tests earlier to fix ambiguity with function definitions.

**Comment by jrappen** [comment]
Are we not sticking to syncing branch_point names with the names of the contexts containing them?

Maybe you could replace images with code blocks in the opening comment.

**Comment by deathaxe** [comment]
If we create a dedicated context with a single pattern forming a branch_point, we likely want to do so.

In this case the patterns forming branch_points are one of many in their context, because it is the one and only location of usage. There could likely be multiple branch_point forming patterns each of which must have a unique name and thus can't equal the context name.

The names have been carefully choosen, except the `cmd-` prefix probably not being needed.

IF those 2 patterns would be organized in a dedicated context, it would be named like the branch_points. But there are no reasons to do so.

Current implementation is a short form of

```diff
@@ -714,6 +714,9 @@
       captures:
         1: punctuation.definition.parameter.shell
       push: cmd-compound-test-pattern
+    - include: cmd-compound-test-comparison
+
+  cmd-compound-test-comparison:
     # comparison expressions
     - match: '{{word_begin}}'
       branch_point: cmd-compound-test-comparison
```

which was avoided due to unique usage of cmd-compound-test-comparison pattern.

**Review by keith-hall — APPROVED** [review]
